draft-pda-protocol.md

```diff
@@ -522,78 +523,69 @@ example, that clients are prohibited from talking to helpers but not the leader.
-  PAHelperInputShare helper_input_share;
-  PALeaderInputShare leader_input_share;
+  PATask task; // Equal to PAUploadStartReq.task
+  uint64 time; // UNIX time (in seconds).
+  PAEncryptedInputShare encrypted_input_shares<1..2^16-1>;
 } PAUploadFinishReq;
 ~~~
 
 We sometimes refer to this message as the *report*. The message contains the
-`task` fields of the previous request. In addition, it includes the time (in
-seconds since the beginning of UNIX time) at which the report was generated, the
-helper's HPKE config id, and the helper and leader shares. The helper share has
-the following structure:
+`task` fields of the previous request. It also includes the time (in seconds
+since the beginning of UNIX time) at which the report was generated. The rest of
+the message consists of the encrypted input shares, each of which has the
+following structure:
 
 ~~~
 struct {
+  uint8 config_id;
   opaque enc<1..2^16-1>;
-  PAProto proto;
-  select (PAHelperInputShare.proto) {
-    case prio: PrioHelperInputShare;
-    case hits: HitsHelperInputShare;
-  }
-} PAHelperInputShare;
+  opaque payload<1..2^16-1>;
+} PAEncryptedInputShare;
+~~~
+
+* `config_id` is equal to `HpkeConfig.id`, where `HpkeConfig` is the
+  aggregator's key config.
+* `enc` is the encapsulated HPKE context, used by the aggregator to decrypt its
+  input share.
+* `payload` is the encrypted input share.
+
+To generate the report, the client begins by encoding its measurements as an
+input for the PA protocol and splitting it into input shares. (Note that the
+structure of each input share depends on the PA protocol in use, its parameters,
+and the role of aggregator, i.e., whether the aggregator is a leader or helper.)
+To encrypt an input share, the client first generates an HPKE
+{{!I-D.irtf-cfrg-hpke}} context for the aggregator by running
+
+~~~
+enc, context = SetupBaseS(pk, [TODO])
 ~~~
 
-Field `enc` encodes the helper's encapsulated HPKE context. The remainder of the
-structure contains the share itself, the structure of which is specific to the
-PA protocol. The structure of the leader share is similarly protocol specific:
+where `pk` is the aggregator's public key. `enc` is the encapsulate HPKE
+context and `context` is the HPKE context used by the client for encryption.
+The payload is encrypted as
 
 ~~~
-struct {
-  PAProto proto;
-  select (PALeaderInputShare.proto) {
-    case prio: PrioLeaderInputShare;
-    case hits: HitsLeaderInputShare;
-  };
-} PALeaderInputShare;
+payload = context.Seal(input_share, [TODO])
 ~~~
 
-Note that only the helper's share is encrypted, and that the leader share is
-sent in plaintext. (Confidentiality of the leader's share is provided by the
-server-authenticated secure channel over which the request is made.)
+where `input_share` is the aggregator's input share.
+
+[TODO: Fully specify encryption of the shares. We need to make sure we
+authenticate things like the PA parameters and the report timestamp. We need to
```

Comment:
I filed #75 to follow up on this, but: what good is authenticating the timestamp if the client can just lie about it?

Reply:
Authenticating the timestamp is for the client's sake. It's meant to ensure its report is only used for a single batch.

Reply:
Hmm, I don't really see how it helps the client at all. Can you elaborate?

Reply:
The idea was discussed a few weeks ago, but hasn't been fully fleshed out. Basically the PA parameters specify a batch window (batch_start, batch_end). To check that a report should be added to a batch, an aggregator decrypts its input share. Decryption will fail if the reported timestamp doesn't match the timestamp generated by the client. This ensures that an honest aggregator only includes an input share in a batch if the timestamp is in the batch window.

Reply:
I get that, but how does this benefit the _client_? (We should just take this to #75.)

Reply:
Done.